# Linear Regression Model


## Objectives

-Load a dataset and preprocess it.

-Train a linear regression model using scikit-learn.

-Interpret the model coefficients.

-Evaluate the model using R-squared and mean squared-error (MSE).

-Tools: Python, pandas, scikit-learn.

In [1]:
# Import Libraries

import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

In [5]:
# Load the dataset

# The file is whitespace-separated, not comma-separated.
columns = [
    "crime_rate",            # CRIM: Per capita crime rate
    "zoned_residential",     # ZN: Large residential lots percentage
    "industrial_prop",       # INDUS: Non-retail business acres percentage
    "near_river",            # CHAS: 1 if bounds Charles River, 0 otherwise
    "nox_concentration",     # NOX: Nitric oxides concentration
    "avg_rooms",             # RM: Average rooms per dwelling
    "prop_old_houses",       # AGE: Built before 1940
    "dist_employment",       # DIS: Distance to employment centers
    "highway_access",        # RAD: Index of highway accessibility
    "tax_rate",              # TAX: Property tax rate per $10,000
    "pupil_teacher_ratio",   # PTRATIO: School pupil-teacher ratio
    "demographic_index",     # B: Historical socio-economic metric
    "lower_status_pct",      # LSTAT: % lower status of population
    "median_house_value"     # MEDV: Target value in $1000s
]

df = pd.read_csv("dataset/house Prediction Data Set.csv", sep=r"\s+", header=None, names=columns)

df.head()

,crime_rate,zoned_residential,industrial_prop,near_river,nox_concentration,avg_rooms,prop_old_houses,dist_employment,highway_access,tax_rate,pupil_teacher_ratio,demographic_index,lower_status_pct,median_house_value
0,0.00632,18.0,2.31,0,0.538,6.575,65.2,4.0900,1,296.0,15.3,396.90,4.98,24.0
1,0.02731,0.0,7.07,0,0.469,6.421,78.9,4.9671,2,242.0,17.8,396.90,9.14,21.6
2,0.02729,0.0,7.07,0,0.469,7.185,61.1,4.9671,2,242.0,17.8,392.83,4.03,34.7
3,0.03237,0.0,2.18,0,0.458,6.998,45.8,6.0622,3,222.0,18.7,394.63,2.94,33.4
4,0.06905,0.0,2.18,0,0.458,7.147,54.2,6.0622,3,222.0,18.7,396.90,5.33,36.2


### Observation
The target variable is **median_house_value**, which represents the house price; the other features are the independent variables used to predict it.


In [6]:
# Checking the info of the data

df.info()

<class 'pandas.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   crime_rate           506 non-null    float64
 1   zoned_residential    506 non-null    float64
 2   industrial_prop      506 non-null    float64
 3   near_river           506 non-null    int64  
 4   nox_concentration    506 non-null    float64
 5   avg_rooms            506 non-null    float64
 6   prop_old_houses      506 non-null    float64
 7   dist_employment      506 non-null    float64
 8   highway_access       506 non-null    int64  
 9   tax_rate             506 non-null    float64
 10  pupil_teacher_ratio  506 non-null    float64
 11  demographic_index    506 non-null    float64
 12  lower_status_pct     506 non-null    float64
 13  median_house_value   506 non-null    float64
dtypes: float64(12), int64(2)
memory usage: 55.5 KB


### Observations:

- There are a total of **506** non-null observations in each of the columns. This indicates that there are no missing values in the data.
- There are **13** columns in the dataset and every column is of numeric data type.

In [7]:
# Exploratory Data Analysis and Data Preprocessing

df.describe().T

,count,mean,std,min,25%,50%,75%,max
crime_rate,506.0,3.613524,8.601545,0.00632,0.082045,0.25651,3.677083,88.9762
zoned_residential,506.0,11.363636,23.322453,0.00000,0.000000,0.00000,12.500000,100.0000
industrial_prop,506.0,11.136779,6.860353,0.46000,5.190000,9.69000,18.100000,27.7400
near_river,506.0,0.069170,0.253994,0.00000,0.000000,0.00000,0.000000,1.0000
nox_concentration,506.0,0.554695,0.115878,0.38500,0.449000,0.53800,0.624000,0.8710
avg_rooms,506.0,6.284634,0.702617,3.56100,5.885500,6.20850,6.623500,8.7800
prop_old_houses,506.0,68.574901,28.148861,2.90000,45.025000,77.50000,94.075000,100.0000
dist_employment,506.0,3.795043,2.105710,1.12960,2.100175,3.20745,5.188425,12.1265
highway_access,506.0,9.549407,8.707259,1.00000,4.000000,5.00000,24.000000,24.0000
tax_rate,506.0,408.237154,168.537116,187.00000,279.000000,330.00000,666.000000,711.0000


### Split the dataset

In [8]:
# Split the dataset into features and target variable
X = df[['avg_rooms']]
y = df[['median_house_value']]

# Split the dataset into training and testing sets

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

#  Handle missing values

imputer = SimpleImputer(strategy="mean")

X_train = imputer.fit_transform(X_train)
X_test = imputer.transform(X_test)

# Train the Linear Regression model
model = LinearRegression()
model.fit(X_train, y_train)

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model
mse = mean_squared_error(y_test, y_pred)
r2 = r2_score(y_test, y_pred)


print("Mean Squared Error:", mse)
print("R-squared Score:", r2)


Mean Squared Error: 46.144775347317264
R-squared Score: 0.3707569232254778


In [9]:
print("Intercept:", model.intercept_)
print("Coefficient:", model.coef_[0])

Intercept: [-36.2463189]
Coefficient: [9.34830141]


### Model Equation
- Recall that **y = mx + c**, therefore the model equation is:
 `median_house_value` = -36.25 + 9.35 × avg_rooms
- Meaning: 
    For every 1 extra room, the predicted house price increases by about 9.35 units.

- The R-squared value is about: 
    **0.37**.
    This means the number of rooms explains about 37% of the variation in house prices.